### PR TITLE
HTML Learning Module in IRSS

### PR DESCRIPTION
**Pull request by chfsx**
image

This PR contains the implementation of the Container Resource GUI as decided yesterday: [IRSS: Container-Resource User-interface](https://docu.ilias.de/goto_docu_wiki_wpage_8127_1357.html)

The changes to the IRSS are in the following commit for your info, which I will cherry-pick in the next few days: https://github.com/ILIAS-eLearning/ILIAS/pull/7214/commits/cdbb7bcec5c8149bf54d76938c926380bb2b4a52

The following commit also contains a readme on how to use the container GUI in your component: https://github.com/ILIAS-eLearning/ILIAS/pull/7214/commits/93e0d7884cf83eccc1815ec1240943ebea6c2b52

as a first use I have already rebuilt the HTML learning module, incl. migration. after I have merged the other two commits, I would give you Alex the PR regarding the changes in the learning module: https://github.com/ILIAS-eLearning/ILIAS/pull/7214/commits/d3c735810ab2e851f9be16ca6d1792ae3d86bbf8

The adjustments to the breadcrumb will follow separately as PR

### CODE REVIEW

**Comment by matthiaskunkel** [comment]
Jour Fixe, 15 APR 2024: @chfsx notified us about the extension of the IRSS to handle multiple files and the first use of it in for the HTML LM. PR will be handled by @alex40724. Please check how Lucene can write an index of the HTML pages in the new structure.

**Comment by alex40724** [comment]
@chfsx Thanks for the PR! I just tried to test the basics. I was able to

- create an HTML learning module
- add files to the main directory
- declare an HTML file as start file (note: in current releases other file types may be start file, too)

But I was not able to open the HTML LM in presentation mode. I tried two objects and always get:

Diese localhost-Seite wurde nicht gefundenFür folgende Webadresse wurde keine Webseite gefunden: http://localhost/trunk/ilias/public/deliver.php/JY3LCsIwEAD_JWdx89ikiZR-yV6SdKPF2pY2PkD8dyseB4aZt1jESRDMSyXoY40Edb1P14mfBMM4xI1gq_Maz0xQtocmUMkReGl3vyBBYxldQm90n3M0ujivnA279w-Kg5j2xTD1_Dpe6m0Un7btOtUoqyR6KX-EqfgoGYsLWRv2rAM3ycmSFKJMNmMK2oT8BQ/-/index.html.

I am not sure if this is related to the component revision. Could you rebase your branch and check, if presentation is ok? Thanks.

File: components/ILIAS/HTMLLearningModule/classes/class.ilObjFileBasedLM.php [hunk 1]
```diff
@@ -33,62 +41,103 @@ public function __construct(
...
 
-        $ilDB->manipulate("INSERT INTO file_based_lm (id, startfile) VALUES " .
-            " (" . $ilDB->quote($this->getId(), "integer") . "," .
-            $ilDB->quote($this->getStartFile(), "text") . ")");
+        // create empty container resource. empty zips are not allowed, we need at least one file which is hidden
+        $empty_zip = $this->archives->zip(
+            [
+                '.info' => Streams::ofString(''),
+            ]
+        );
```
Comment (alex40724) [inline components/ILIAS/HTMLLearningModule/classes/class.ilObjFileBasedLM.php:88]: I understand the technical reason. But is this something consumer code needs "to know" and take care of?

**Review by alex40724 — CHANGES_REQUESTED** [review]
@chfsx  Any chance of getting a rebased version that fixes the presentation of the learning modules? I would love to merge this.

**Comment by chfsx** [comment]
@alex40724 the PR has now been rebased. If you still face issues aka "Diese localhost-Seite wurde nicht gefundenFür folgende Webadresse wurde keine Webseite gefunden:" we should meet using Discord/Skype/Whatever since I cannot reproduce this. but i remember that you sometimes have very low limits in the OS for your development machines, i know of one case where this also led to problems with the signed URLs. but we could look into that, basically there is also a solution for that. 

Concerning the "empty" zip: I changed that to internal and adapted the code in `ilObjFileBasedLM`

**Comment by alex40724** [comment]
Are you able to run composer du in this branch?

I tried the first time I ran into the missing dependency issues.

I out-commented the the missing dependencies (lots of them), but still I am not able to test this, since I cannot add new objects in this branch.

@chfsx Does this work for you?

**Comment by chfsx** [comment]
of course I can't solve the general issues in trunk with this PR, I have an installation with existing HTML learning modules, which I update to test this. the creation of new objects is generally broken in trunk FMPOV.

**Comment by alex40724** [comment]
@chfsx The modals in the trunk are working again. I created a local branch with your commit. For reasons I do not understand the DB Update does not run, so I added the rid field manually.

I then ran into the same issue getting a "Für folgende Webadresse wurde keine Webseite gefunden..." error.

I tracked this down to this line
```
169            $file_inside_zip_uri = "zip://$requested_zip#$sub_request";
```
in StreamDelivery.

After adding an additional "/" after the "#" it works.
```
-->            $file_inside_zip_uri = "zip://$requested_zip#/$sub_request";
```

Does this work for you, too? No idea, if this is related to the zip lib version.

**Comment by alex40724** [comment]
@chfsx  Since this is not related to the files included here, I merged this manually to the trunk now.

For the delivery issue I added a mantis report. https://mantis.ilias.de/view.php?id=42062

Thanks!

**Pull request closed without merging**